modules/sdk-core/src/bitgo/tss/common.ts

```diff
@@ -68,3 +70,51 @@ export async function sendSignatureShare(
     })
     .result();
 }
+
+/**
+ * Verifies that a TSS wallet signature was produced with the expected key and that the signed data contains the
+ * expected common keychain as well as the expected user and backup key ids
+ */
+export async function commonVerifyWalletSignature(params: {
+  walletSignature: openpgp.Key;
+  bitgoPub: openpgp.Key;
+  commonKeychain: string;
+  userKeyId: string;
+  backupKeyId: string;
+}): Promise<{ value: ArrayBuffer }[]> {
+  const { walletSignature, bitgoPub, commonKeychain, userKeyId, backupKeyId } = params;
+
+  // The result of verifyPrimaryUser contains two items:
+  // * the self-cert of the CA key
+  // * the cert of the subkey (our proof that we are looking for)
+  // By comparing the fingerprints of the walletSignature and the passed bitgoPub we ensure that the passed bitgoPub is
+  // not the CA key (if they are the same, then it is).
+  if (walletSignature.keyPacket.getFingerprint() === bitgoPub.keyPacket.getFingerprint()) {
+    throw new Error('Invalid HSM GPG signature');
+  }
+
+  const verificationResult = await walletSignature.verifyPrimaryUser([bitgoPub]);
+  const isValid = verificationResult.some((result) => result.valid);
+  if (!isValid) {
+    throw new Error('Invalid HSM GPG signature');
+  }
+  const primaryUser = await walletSignature.getPrimaryUser();
+
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore the rawNotations property is missing from the type but it actually exists
+  const rawNotations: { value: Uint8Array }[] = primaryUser.user.otherCertifications[0].rawNotations;
+
+  assert(rawNotations.length === 5, 'invalid wallet signatures');
+
+  assert(
+    commonKeychain === Buffer.from(rawNotations[0].value).toString(),
+    'wallet signature does not match common keychain'
+  );
+  assert(userKeyId === Buffer.from(rawNotations[1].value).toString(), `wallet signature does not match user key id`);
+  assert(
+    backupKeyId === Buffer.from(rawNotations[2].value).toString(),
+    'wallet signature does not match backup key id'
+  );
+
```

Comment:
what about the `userSharePub` and `backupSharePub`?

Remember, from the HSM, there are no uSig proofs.   Only wallet sigs.    which are just public keys of the user and backup party, with attached certifications from the HSM pubkey.   And each of those certs has notation data attached.

Reply:
The motivation was that we can reuse this function when we implement almost the same for ecdsa. As far as I understand it the code in this common function will remain the same but checking the user/backup share pubs will be different depending on the curve.

Reply:
ah, fair point.   thanks for the explanation.